### PR TITLE
Inject user identifier to ExecutionSpec

### PR DESCRIPTION
**Pull request by ByronHsu**
# TL;DR

This pr provides a middleware to inject user identifier to ExecutionSpec.
By default, the value of the user identifier is userid from access/id token.
Users can customize their own middleware and inject different values.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3566

### CODE REVIEW

File: auth/identity_context.go [hunk 1]
```diff
@@ -28,6 +28,8 @@ type IdentityContext struct {
 	appID           string
 	authenticatedAt time.Time
 	userInfo        *service.UserInfoResponse
+	// Contains unique identifier for users, which will be passed to SecurityContext
```
Comment (wild-endeavor) [inline auth/identity_context.go:31]: add more comment explaining why this is separate?

File: auth/identity_context.go [hunk 2]
```diff
@@ -58,6 +60,10 @@ func (c IdentityContext) IsEmpty() bool {
 	return c == emptyIdentityContext
 }
 
+func (c IdentityContext) UserIdentity() string {
+	return c.userIdentity
```
Comment (wild-endeavor) [inline auth/identity_context.go:64]: is this supposed to be `userIdentifier`?

File: auth/identity_context.go [hunk 3]
```diff
@@ -81,6 +86,16 @@ func (c IdentityContext) AuthenticatedAt() time.Time {
 	return c.authenticatedAt
 }
 
+func (c IdentityContext) UserIdentifier() string {
+	return c.userIdentifier
+}
+
+// SetUserIdentifier allows you to explicitly set user identifier
+func (c *IdentityContext) SetUserIdentifier(id string) IdentityContext {
```
Comment (wild-endeavor) [inline auth/identity_context.go:94]: does this need to return something?  can it just not return anything?

File: pkg/manager/impl/execution_manager.go [hunk 4]
```diff
@@ -403,6 +403,18 @@ func (m *ExecutionManager) getExecutionConfig(ctx context.Context, request *admi
...
+
+	if workflowExecConfig.GetSecurityContext().GetRunAs() == nil {
+		workflowExecConfig.SecurityContext.RunAs = &core.Identity{}
+	}
+
+	// In the case of reference_launch_plan subworkflow, the context comes from flytepropeller instead of the user side, so user auth is missing.
+	// We skip getUserIdentityFromContext but can still get ExecUserId because flytepropeller passes it in the execution request.
+	// https://github.com/flyteorg/flytepropeller/blob/03a6672960ed04e7687ba4f790fee9a02a4057fb/pkg/controller/nodes/subworkflow/launchplan/admin.go#L114
+	if workflowExecConfig.GetSecurityContext().GetRunAs().GetUserIdentifier() == "" {
+		workflowExecConfig.SecurityContext.RunAs.UserIdentifier = auth.IdentityContextFromContext(ctx).UserIdentifier()
```
Comment (wild-endeavor) [inline pkg/manager/impl/execution_manager.go:415]: when admin is running without auth, this still won't fail right?

Comment (ByronHsu) [reply pkg/manager/impl/execution_manager.go:415]: yeah UserIdentifier would just be an empty string

File: auth/identity_context.go (hunk 3 above)
Comment (ByronHsu) [reply auth/identity_context.go]: changed

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]

File: auth/identity_context.go [hunk 5]
```diff
@@ -81,6 +86,15 @@ func (c IdentityContext) AuthenticatedAt() time.Time {
 	return c.authenticatedAt
 }
 
+func (c IdentityContext) UserIdentifier() string {
+	return c.userIdentifier
+}
```
Comment (EngHabu) [inline auth/identity_context.go:91]: This is confusing when we have `UserID()` can you more explicitly name this?

Comment (wild-endeavor) [reply auth/identity_context.go:91]: how about `executionUserIdentifier`

**Review by wild-endeavor — APPROVED** [review]

**Review by pingsutw — DISMISSED** [review]

**Review by wild-endeavor — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]
